### PR TITLE
Disable update of proxy rule position when account is gone

### PR DESCRIPTION
**Pull request by guicassolato**
Deletion of multiple proxy rule objects in parallel, which may happen when an Account or a Backend is being deleted in the background, can deadlock on the update of the "position" of the proxy rules. This PR disables `acts_as_list` position update when a proxy rule is destroyed if (and only if) the proxy rule's owner or the account top of it has been deleted (scheduled for deletion).

Closes [THREESCALE-3845](https://issues.jboss.org/browse/THREESCALE-3845)

### CODE REVIEW

File: app/models/proxy_rule.rb [hunk 1]
```diff
@@ -124,6 +124,18 @@ def backend_api_owner?
...
+
+  def destroy
+    if !owner || !account || owner.scheduled_for_deletion?
+      self.class.acts_as_list_no_update do
+        super
+      end
+    else
+      super
+    end
+  end
```
Comment (guicassolato) [inline app/models/proxy_rule.rb]: A couple alternatives:

1.
```ruby
  def destroy
    return super if owner&.persisted? && account&.persisted? && !owner.scheduled_for_deletion?

    acts_as_list_no_update do
      super
    end
  end
```

2.
```ruby
  def scheduled_for_deletion?
    !owner || !account || owner.scheduled_for_deletion?
  end

  def act_as_list_no_update?
    super || scheduled_for_deletion?
  end
```

File: test/unit/proxy_rule_test.rb [hunk 2]
```diff
@@ -157,6 +157,46 @@ class ProxyRuleTest < ActiveSupport::TestCase
...
+
+    attr_reader :provider
+
+    test 'position is not updated when account has been deleted' do
+      proxy = FactoryBot.create(:service, account: provider).proxy
+      proxy_rule = FactoryBot.create(:proxy_rule, proxy: proxy)
+
+      provider.delete
+
+      proxy_rule.expects(:decrement_positions_on_lower_items).never
```
Comment (guicassolato) [inline test/unit/proxy_rule_test.rb]: This is the downside of this approach using `#act_as_list_no_update?`.

File: app/events/proxy_configs/affecting_object_changed_event.rb [hunk 3]
```diff
@@ -8,8 +8,12 @@ def self.create(proxy, object)
       object_type: object.class.name,
       metadata: {
         service_id: proxy.service_id,
-        provider_id: proxy.account.id
+        provider_id: proxy.account&.id
```
Comment (hallelujah) [inline app/events/proxy_configs/affecting_object_changed_event.rb]: Won’t the event be invalid in this case?

**Comment by hallelujah** [comment]
Is the update position inside a callback?
If so it can be skipped

File: app/models/proxy_rule.rb [hunk 4]
```diff
@@ -124,6 +124,16 @@ def backend_api_owner?
     owner_type == 'BackendApi'
   end
 
+  delegate :account, to: :owner, allow_nil: true
+
+  def scheduled_for_deletion?
+    !owner || !account || owner.scheduled_for_deletion?
+  end
+
+  def act_as_list_no_update?
```
Comment (hallelujah) [inline app/models/proxy_rule.rb]: OK this is the callback condition

File: test/unit/proxy_rule_test.rb (hunk 2 above)
Comment (hallelujah) [reply test/unit/proxy_rule_test.rb]: I think it is fine
The proxy_rule is going to be deleted anyway no?
Unless you remove the `!account` In scheduled_for_position

File: app/events/proxy_configs/affecting_object_changed_event.rb (hunk 3 above)
Comment (guicassolato) [reply app/events/proxy_configs/affecting_object_changed_event.rb]: Second time this happens to me in less than a month. First add the nil type check, then define the `.valid?` class method and forget to remove the check. Sorry :/

File: app/events/proxy_configs/affecting_object_changed_event.rb [hunk 5]
```diff
@@ -12,4 +12,8 @@ def self.create(proxy, object)
       }
     )
   end
+
+  def self.valid?(proxy, *_args)
+    proxy && proxy.account&.persisted? # Proxy's service or account may have been deleted
```
Comment (hallelujah) [inline app/events/proxy_configs/affecting_object_changed_event.rb]: Good for me :)

It is more explicit that this event can't be valid if account is not there anymore

Though it reeks code smell by the use of `account&.` The reason is simple, do we generate config change on deletion?

Comment (guicassolato) [reply app/events/proxy_configs/affecting_object_changed_event.rb]: We do issue Affecting Object Changes events on deletion, yes.

Comment (hallelujah) [reply app/events/proxy_configs/affecting_object_changed_event.rb]: Ouch :(
Can't it be fixed when it is scheduled for deletion or maybe?
For the issue you address the PR is good enough but it looks like a can of worm if we want to fix everything.
Good for you to address (if needed) the above issue in another PR?

Comment (guicassolato) [reply app/events/proxy_configs/affecting_object_changed_event.rb]: @hallelujah, thinking about moving the check here to https://github.com/3scale/porta/blob/1e8bdf5d3ab494f09d2db7e471f690860f32b98f/app/lib/proxy_config_affecting_changes.rb#L13?

Comment (guicassolato) [reply app/events/proxy_configs/affecting_object_changed_event.rb]: https://issues.redhat.com/browse/THREESCALE-4322

**Pull request merged**